ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -265,8 +588,26 @@ module AuthADWinbind : Auth_signature.AUTH_MODULE = struct
       it's a string*string list anyway for possible future expansion.
       Raises Not_found (*Subject_cannot_be_resolved*) if subject_id cannot be resolved by external auth service
   *)
-  let query_subject_information subject_identifier =
-    failwith "extauth_plugin authenticate_ticket not implemented"
+  let query_subject_information (sid : string) =
+    (* we have to aggregate information from wbinfo and / or an ldap query (using net binary)
+     * to begin with, we don't even know if the sid belongs to a group or a user
+     * so we first do a test with [gid_of_sid] + [uid_of_sid] *)
+    let res =
+      let* name = Wbinfo.name_of_sid sid in
+      match Wbinfo.uid_of_sid sid with
+      | Ok uid ->
```

Comment:
why need to use `uid` as identifier for the user here, other than `SID` directly?
`uid` may not right (consistent) and may not trusted as 
- hcp_nss will override the uid to `0`, to force all users login as root, we only support one single root in dom0
- pbis and users have different uid for the same domain user, for upgrade case, we cannot keep it consistent.

Reply:
I have re-consider this, currently implementation probably good enough as
- hcp_nss only override uid to 0 during resolve uid like `getpwnam_r`, here we ask wbinfo `winbind` for the UID directly, so it is not overrided.
- The uid here is only used temporary, for the upgraded case, the UID is also regenerated by winbind and used to query winbind info, so it should be fine.

So, currently implementation looks good to me,  Just for your info, if you do want to distinguish a user/group from SID, you can use  `wbinfo -s, --sid-to-name=SID,`, and it will tell you it is a group or user, 1 for user, 2 for group, like

```[root@xrtuk-14-05 test_nss]# wbinfo -s S-1-5-21-3143668282-2591278241-912959342-1174
CONNAPP\test_group 2
[root@xrtuk-14-05 test_nss]# wbinfo -s S-1-5-21-3143668282-2591278241-912959342-1178
CONNAPP\ladmin 1```

Reply:
We need to get the uid, since it's one of the fields returned when you call `query_subject_information`

> pbis and users have different uid for the same domain user, for upgrade case, we cannot keep it consistent.

true, after the upgrade we'll need to make sure that any subject rows in the database are updated accordingly :+1: 

> and it will tell you it is a group or user, 1 for user, 2 for group

thanks, I will add that simplification!